tehuti-vis.py

```diff
@@ -17,78 +17,95 @@
+        self.plot_data = None
+
+        self.methods = {'basic': VaryRepoCommit(self),
+                        'violin': Violin(self)}
+
+    @property
+    def method(self):
+        if self._method is None:
+            self._method = 'basic'
+        return self.methods[self._method]
+
+    @method.setter
+    def method(self, value):
         try:
-            k = gaussian_kde(d) #calculates the kernel density
-        except ValueError:
-            ax.hlines(d, p-w, p+w)
-        else:
-            line = False
-            m = k.dataset.min() #lower bound of violin
-            M = k.dataset.max() #upper bound of violin
-            x = arange(m,M,(M-m)/100.) # support for violin
-            v = k.evaluate(x) #violin profile (density curve)
-            v = v/v.max()*w #scaling the violin to the available space
-            ax.fill_betweenx(x,p,v+p,facecolor='y',alpha=0.3)
-            ax.fill_betweenx(x,p,-v+p,facecolor='y',alpha=0.3)
-    if bp:
-        ax.boxplot(data,notch=1,positions=pos,vert=1)
-
-
-def plot(picked, single_id):
-    keys = None
-    for commit, results in picked:
-        if keys is None:
-            keys = results.viewkeys()
-        else:
-            keys = keys & results.viewkeys()
-    for key in keys:
-        if key == 'name' or (single_id and key != single_id):
-            continue
-        labels = []
-        data = []
-        for commit, results in picked:
-            labels.append(shorten_sha(commit))
-            data.append(results[key])
-        ax = plt.axes()
-        ax.set_title(key)
-        if isinstance(data[0], list):
-            violin_plot(ax, data)
-        else:
-            width = 0.6
-            pos = [x - (width / 2) for x in range(len(data))]
-            ax.bar(pos, data, width=width)
-        ax.set_xticks(range(len(data)))
-        ax.set_xticklabels(labels, rotation=30)
-        plt.tight_layout()
-        plt.show()
+            self._method = self.methods[value]
+        except KeyError:
+            msg = 'Vis has no method {}.'
+            raise AttributeError(msg.format(value))
+
+    def select_data(self, commits=None, metrics=None):
+        """
+        Select data to plot. Processing is handed off to the `select_data`
+        method of the current state.
+
+        Args:
+
+         * commits:
+            One or more valid repository commits that have been benchmarked.
+         * metrics:
+            One or more metrics that have results in the metrics results file.
+
+        """
+        self.plot_data = self.method.select_data(commits, metrics)
+
+    def plot(self, alternate_plot=False):
+        """
+        Plot data. Plotting is handed off to the `plot` method of the current
+        state.
+
+        Arg:
+
+        * alternate_plot:
+            Toggle to select the alternate plotting mode of the `plot` method
+            of the current state.
+
+        """
+        if self.plot_data is None:
+            self.select_data()
```

Comment:
Yes, `self.select_data()` should always populate `self.plot_data`, but it doesn't _have_ to be called (though in practice it is likely to be as this is all designed to be run from the `__main__` section of this module).
Not setting the kwargs for `self.select_data()` actually means that all metrics and commits found in the results file are selected to be plotted, but this wasn't clear from it's docstring. I've updated the docstring to make this clearer.